### PR TITLE
[FEAT] Visit other Farms

### PR DESCRIPTION
**Pull request by j923sneaks**
# Description

This PR implements visiting other farms by searching their Farm ID. If found, it will load the balance and inventory of said farm. The fields are filled with Sunflowers for decorative purposes; also since we cannot get session data of farm via web3 calls.

Other features:
- Removed Hud parts
- Non clickable fields, blacksmith, market, well
- Back button goes back to connection state

Note that I have not implemented searching of farm via URL query due to the ff reasons:
- I think it will be tricky managing states if the URL query is always present (i.e. back button may trigger search again)
- Honestly haven not thought which state this could be attached
- Can the query be check in `invoke src` instead of `cond`?

Nonetheless, I encourage anyone to add the query on a separate PR 😄 

Fixes/Implements #258 (Partial)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Manual Testing. I used Farm ID 1 for these tests. Ignored dirty data as long as they are correctly displayed.
sfl-visit-farm-feat

Account 1 -- has existing farm

https://user-images.githubusercontent.com/89294757/154088364-891315e8-ac45-445d-a013-5e486633b410.mp4

Account 2 -- no existing farm

https://user-images.githubusercontent.com/89294757/154088427-81b6ad94-65cd-4664-873c-6877f450e5dc.mp4

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/lib/blockchain/Inventory.ts [hunk 1]
```diff
@@ -43,4 +45,21 @@ export class Inventory {
       {} as ItemSupply
     );
   }
+
+  public async getBalances(farmAddress: string) {
+    const batchAccounts = Array(IDS.length).fill(farmAddress);
+    const balances = await this.contract.methods.balanceOfBatch(batchAccounts, IDS).call();
+    const names = Object.keys(KNOWN_IDS) as InventoryItemName[];
+
+    return balances.reduce((items: Inventory, balance: string, index: number) => {
```
Comment (adamhannigan) [inline src/lib/blockchain/Inventory.ts:54]: Similar to above, I would suggest moving the game logic out of this Inventory class and into a separate helper class.

That helps make this file focussed on Blockchain interacts and also makes it easier to test the balances -> Inventory code.

E.g
```
const balances = await metamask.getInventory().getBalances(farmAddress);
const inventory = balancesToInventory(balances) // where the code would live
```

File: src/features/game/lib/gameMachine.ts [hunk 2]
```diff
@@ -123,14 +122,18 @@ export function startGame(authContext: AuthContext) {
...
+            // Visit farm 
+            const address = authContext.address as string;
+            const game = await getVisitState(address);
 
-            return { state: INITIAL_FARM };
+            game.id = authContext.farmId as number;
+
+            return { state: game };
+
+            // return { state: INITIAL_FARM };
```
Comment (adamhannigan) [inline src/features/game/lib/gameMachine.ts:133]: We also have the case where the `API_URL` environment variable is null. This is used for quickly running the repo and on the `https://sunflower-land.com/art-preview-only` URL.

You will want to consider how we can support those cases. Right now it throws an error

Screen Shot 2022-02-15 at 2 24 38 pm

**Review by adamhannigan — CHANGES_REQUESTED** [review]
Awesome job 👏 It makes sense for someone to do the query in a separate PR as well.

I have left a few suggestions but overall the code looks good 💪 

There is one small bug that if no `VITE_API_URL` is provided, no game state is loaded. We will want to fix this to assist in quick UI development :)

File: src/lib/blockchain/Farm.ts [hunk 3]
```diff
@@ -36,4 +36,10 @@ export class Farm {
 
     return accounts;
   }
+
+  public async visitFarm(tokenId: number): Promise<FarmAccount> {
```
[the remark this answers is not in the document]
Comment (j923sneaks) [reply src/lib/blockchain/Farm.ts]: renamed to `getFarm` and now just returns FarmAccount

File: src/lib/blockchain/Inventory.ts (hunk 1 above)
Comment (j923sneaks) [reply src/lib/blockchain/Inventory.ts]: created separate file `visitUtils.ts` and put `balancesToInventory()` there

File: src/features/game/lib/gameMachine.ts (hunk 2 above)
Comment (j923sneaks) [reply src/features/game/lib/gameMachine.ts]: Thanks for this one, did not notice 😅 

made changes in `gameMachine.ts` to handle preview mode i.e. returning `INITIAL_FARM`

sfl-visit-farm-initital-farm

File: src/features/game/actions/visit.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { metamask } from "lib/blockchain/metamask";
+import { fromWei } from "web3-utils";
+import Decimal from "decimal.js-light";
+
+import { GameState, FieldItem } from "../types/game";
+
+export async function getVisitState(farmAddress: string): Promise<GameState> {
+  const balance = await metamask.getToken().balanceOf(farmAddress);
+  const fields = Array(22).fill({ name: "Sunflower", plantedAt: 0 });
```
[the remark this answers is not in the document]
Comment (j923sneaks) [reply src/features/game/actions/visit.ts]: added `populateFields()` in `visitUtils.ts` to check inventory items before filling `fields`. Tested in Farm ID 2

sfl-visit-partial-pic

**Comment by j923sneaks** [comment]
hi @adamhannigan thanks for the feedback! i have resolved all comments and also pulled from main to use lint.

**Review by adamhannigan — APPROVED** [review]
Great work @j923sneaks 

This will be awesome for sharing farms 🤗

**Pull request merged**